### ISSUE TITLE
Server: fatal error: concurrent map writes

### ISSUE THREAD

**Issue by djavorszky** [djavorszky/ddn#114]
Server died with the following fatal error:

```
fatal error: concurrent map writes

goroutine 832384 [running]:
runtime.throw(0x8a78c4, 0x15)
	/usr/local/go/src/runtime/panic.go:596 +0x95 fp=0xc450489700 sp=0xc4504896e0
runtime.mapassign(0x825560, 0xc4201380f0, 0xc4204fe0f0, 0x0)
	/usr/local/go/src/runtime/hashmap.go:499 +0x667 fp=0xc4504897a0 sp=0xc450489700
main.register(0xa87ac0, 0xc420282000, 0xc42045ad00)
	/home/javdaniel/go/src/github.com/djavorszky/ddn/server/handlers.go:413 +0x2b2 fp=0xc450489b18 sp=0xc4504897a0
net/http.HandlerFunc.ServeHTTP(0x8c1c78, 0xa87ac0, 0xc420282000, 0xc42045ad00)
	/usr/local/go/src/net/http/server.go:1942 +0x44 fp=0xc450489b40 sp=0xc450489b18
github.com/djavorszky/ddn/common/srv.Logger.func1(0xa87ac0, 0xc420282000, 0xc42045ad00)
	/home/javdaniel/go/src/github.com/djavorszky/ddn/common/srv/srv.go:14 +0x9c fp=0xc450489be8 sp=0xc450489b40
net/http.HandlerFunc.ServeHTTP(0xc420132140, 0xa87ac0, 0xc420282000, 0xc42045ad00)
	/usr/local/go/src/net/http/server.go:1942 +0x44 fp=0xc450489c10 sp=0xc450489be8
github.com/gorilla/mux.(*Router).ServeHTTP(0xc4201160a0, 0xa87ac0, 0xc420282000, 0xc42045ad00)
	/home/javdaniel/go/src/github.com/gorilla/mux/mux.go:114 +0x10c fp=0xc450489d10 sp=0xc450489c10
net/http.serverHandler.ServeHTTP(0xc42019c000, 0xa87ac0, 0xc420282000, 0xc420316600)
	/usr/local/go/src/net/http/server.go:2568 +0x92 fp=0xc450489d58 sp=0xc450489d10
net/http.(*conn).serve(0xc42013d9a0, 0xa88200, 0xc4203dfc40)
	/usr/local/go/src/net/http/server.go:1825 +0x612 fp=0xc450489fc8 sp=0xc450489d58
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc450489fd0 sp=0xc450489fc8
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2668 +0x2ce

goroutine 1 [IO wait]:
net.runtime_pollWait(0x7fea0039dfd8, 0x72, 0xa84900)
	/usr/local/go/src/runtime/netpoll.go:164 +0x59
net.(*pollDesc).wait(0xc420175878, 0x72, 0xa805f8, 0xc420426ee0)
	/usr/local/go/src/net/fd_poll_runtime.go:75 +0x38
net.(*pollDesc).waitRead(0xc420175878, 0xffffffffffffffff, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:80 +0x34
net.(*netFD).accept(0xc420175810, 0x0, 0xa83180, 0xc420426ee0)
	/usr/local/go/src/net/fd_unix.go:430 +0x1e5
net.(*TCPListener).accept(0xc42013e110, 0xc4200c35c0, 0x81aee0, 0xffffffffffffffff)
	/usr/local/go/src/net/tcpsock_posix.go:136 +0x2e
net.(*TCPListener).AcceptTCP(0xc42013e110, 0xc420047a10, 0xc420047a18, 0xc420047a08)
	/usr/local/go/src/net/tcpsock.go:215 +0x49
net/http.tcpKeepAliveListener.Accept(0xc42013e110, 0x8c1eb0, 0xc4200c3540, 0xa882c0, 0xc420139e60)
	/usr/local/go/src/net/http/server.go:3044 +0x2f
net/http.(*Server).Serve(0xc42019c000, 0xa87cc0, 0xc42013e110, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:2643 +0x228
net/http.(*Server).ListenAndServe(0xc42019c000, 0xc42019c000, 0x9)
	/usr/local/go/src/net/http/server.go:2585 +0xb0
net/http.ListenAndServe(0xc420134071, 0x5, 0xa82340, 0xc4201160a0, 0x1, 0xc420134071)
	/usr/local/go/src/net/http/server.go:2787 +0x7f
main.main()
	/home/javdaniel/go/src/github.com/djavorszky/ddn/server/main.go:123 +0x6ea

goroutine 17 [syscall, 10014 minutes, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2197 +0x1

goroutine 5 [syscall, 10014 minutes]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:116 +0x104
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x41

goroutine 6 [select, 10014 minutes, locked to thread]:
runtime.gopark(0x8c2468, 0x0, 0x89d5f6, 0x6, 0x18, 0x2)
	/usr/local/go/src/runtime/proc.go:271 +0x13a
runtime.selectgoImpl(0xc420032f50, 0x0, 0x18)
	/usr/local/go/src/runtime/select.go:423 +0x1364
runtime.selectgo(0xc420032f50)
	/usr/local/go/src/runtime/select.go:238 +0x1c
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal_unix.go:434 +0x2dd
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2197 +0x1

goroutine 7 [chan receive, 10014 minutes]:
main.main.func2(0xc42005e240)
	/home/javdaniel/go/src/github.com/djavorszky/ddn/server/main.go:35 +0x44
created by main.main
	/home/javdaniel/go/src/github.com/djavorszky/ddn/server/main.go:39 +0x154

goroutine 10 [chan receive, 10014 minutes]:
database/sql.(*DB).connectionOpener(0xc4200c2e60)
	/usr/local/go/src/database/sql/sql.go:837 +0x4a
created by database/sql.Open
	/usr/local/go/src/database/sql/sql.go:582 +0x212

goroutine 19 [runnable]:
main.increment(0xc4201380f0)
	/home/javdaniel/go/src/github.com/djavorszky/ddn/server/registry.go:28 +0x56
created by main.initRegistry
	/home/javdaniel/go/src/github.com/djavorszky/ddn/server/registry.go:16 +0x7a

goroutine 20 [chan receive, 1379 minutes]:
main.maintain()
	/home/javdaniel/go/src/github.com/djavorszky/ddn/server/process.go:25 +0xbc
created by main.main
	/home/javdaniel/go/src/github.com/djavorszky/ddn/server/main.go:116 +0x5fa

goroutine 21 [chan receive]:
main.checkConnectors()
	/home/javdaniel/go/src/github.com/djavorszky/ddn/server/process.go:95 +0xb5
created by main.main
	/home/javdaniel/go/src/github.com/djavorszky/ddn/server/main.go:119 +0x612

goroutine 832339 [runnable]:
net/http.(*connReader).backgroundRead(0xc4203dfc80)
	/usr/local/go/src/net/http/server.go:655
created by net/http.(*connReader).startBackgroundRead
	/usr/local/go/src/net/http/server.go:652 +0xdf

goroutine 832359 [runnable]:
net/http.(*conn).serve(0xc4200c34a0, 0xa882c0, 0xc420139e60)
	/usr/local/go/src/net/http/server.go:1715
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2668 +0x2ce

goroutine 827283 [IO wait, 9 minutes]:
net.runtime_pollWait(0x7fea0039d798, 0x72, 0x7)
	/usr/local/go/src/runtime/netpoll.go:164 +0x59
net.(*pollDesc).wait(0xc4204e5bf8, 0x72, 0xa84900, 0xa805f8)
	/usr/local/go/src/net/fd_poll_runtime.go:75 +0x38
net.(*pollDesc).waitRead(0xc4204e5bf8, 0xc420459000, 0x1000)
	/usr/local/go/src/net/fd_poll_runtime.go:80 +0x34
net.(*netFD).Read(0xc4204e5b90, 0xc420459000, 0x1000, 0x1000, 0x0, 0xa84900, 0xa805f8)
	/usr/local/go/src/net/fd_unix.go:250 +0x1b7
net.(*conn).Read(0xc420128138, 0xc420459000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:181 +0x70
net/http.(*connReader).Read(0xc4203bae80, 0xc420459000, 0x1000, 0x1000, 0x3, 0xc4203243c0, 0x46)
	/usr/local/go/src/net/http/server.go:754 +0x140
bufio.(*Reader).fill(0xc420248120)
	/usr/local/go/src/bufio/bufio.go:97 +0x117
bufio.(*Reader).ReadSlice(0xc420248120, 0xa, 0x1e0, 0x3, 0x160, 0xc420329200, 0xc42001c600)
	/usr/local/go/src/bufio/bufio.go:338 +0xbb
bufio.(*Reader).ReadLine(0xc420248120, 0xc4201a7200, 0x100, 0xf8, 0x890720, 0xa, 0x45d172)
	/usr/local/go/src/bufio/bufio.go:367 +0x37
net/textproto.(*Reader).readLineSlice(0xc42036e510, 0xc4204e1ac8, 0xc4204e1ac8, 0x410b98, 0x100, 0x890720)
	/usr/local/go/src/net/textproto/reader.go:55 +0x5f
net/textproto.(*Reader).ReadLine(0xc42036e510, 0xc4201a7200, 0x72, 0x8000000000000000, 0x0)
	/usr/local/go/src/net/textproto/reader.go:36 +0x2f
net/http.readRequest(0xc420248120, 0x0, 0xc4201a7200, 0x0, 0x0)
	/usr/local/go/src/net/http/request.go:918 +0xa5
net/http.(*conn).readRequest(0xc42019ab40, 0xa88200, 0xc4203bad80, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:934 +0x213
net/http.(*conn).serve(0xc42019ab40, 0xa88200, 0xc4203bad80)
	/usr/local/go/src/net/http/server.go:1763 +0x49a
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2668 +0x2ce

goroutine 829599 [IO wait, 9 minutes]:
net.runtime_pollWait(0x7fea0039e098, 0x72, 0x9)
	/usr/local/go/src/runtime/netpoll.go:164 +0x59
net.(*pollDesc).wait(0xc4201741b8, 0x72, 0xa84900, 0xa805f8)
	/usr/local/go/src/net/fd_poll_runtime.go:75 +0x38
net.(*pollDesc).waitRead(0xc4201741b8, 0xc420400000, 0x1000)
	/usr/local/go/src/net/fd_poll_runtime.go:80 +0x34
net.(*netFD).Read(0xc420174150, 0xc420400000, 0x1000, 0x1000, 0x0, 0xa84900, 0xa805f8)
	/usr/local/go/src/net/fd_unix.go:250 +0x1b7
net.(*conn).Read(0xc42000e030, 0xc420400000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:181 +0x70
net/http.(*connReader).Read(0xc420348180, 0xc420400000, 0x1000, 0x1000, 0x3, 0xc4203243c0, 0x3e)
	/usr/local/go/src/net/http/server.go:754 +0x140
bufio.(*Reader).fill(0xc4504ca120)
	/usr/local/go/src/bufio/bufio.go:97 +0x117
bufio.(*Reader).ReadSlice(0xc4504ca120, 0xa, 0x1e0, 0x3, 0x200, 0xc4200c5980, 0xc42001a000)
	/usr/local/go/src/bufio/bufio.go:338 +0xbb
bufio.(*Reader).ReadLine(0xc4504ca120, 0xc4504abf00, 0x100, 0xf8, 0x890720, 0xa, 0x45d172)
	/usr/local/go/src/bufio/bufio.go:367 +0x37
net/textproto.(*Reader).readLineSlice(0xc4202e4240, 0xc4201ffac8, 0xc4201ffac8, 0x410b98, 0x100, 0x890720)
	/usr/local/go/src/net/textproto/reader.go:55 +0x5f
net/textproto.(*Reader).ReadLine(0xc4202e4240, 0xc4504abf00, 0x72, 0x8000000000000000, 0x0)
	/usr/local/go/src/net/textproto/reader.go:36 +0x2f
net/http.readRequest(0xc4504ca120, 0x0, 0xc4504abf00, 0x0, 0x0)
	/usr/local/go/src/net/http/request.go:918 +0xa5
net/http.(*conn).readRequest(0xc4204fe000, 0xa88200, 0xc420348100, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:934 +0x213
net/http.(*conn).serve(0xc4204fe000, 0xa88200, 0xc420348100)
	/usr/local/go/src/net/http/server.go:1763 +0x49a
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2668 +0x2ce

goroutine 832361 [runnable]:
net/http.(*connReader).backgroundRead(0xc4203f16c0)
	/usr/local/go/src/net/http/server.go:655
created by net/http.(*connReader).startBackgroundRead
	/usr/local/go/src/net/http/server.go:652 +0xdf

goroutine 832360 [runnable]:
reflect.ValueOf(0x88c8a0, 0xc4203d4b40, 0xc420395600, 0x4295d6, 0x0)
	/usr/local/go/src/reflect/value.go:2095 +0x117
fmt.(*pp).printArg(0xc420329200, 0x88c8a0, 0xc4203d4b40, 0xc400000076)
	/usr/local/go/src/fmt/print.go:682 +0x1d4
fmt.(*pp).doPrintf(0xc420329200, 0x89b1e7, 0x3, 0xc420395828, 0x1, 0x1)
	/usr/local/go/src/fmt/print.go:998 +0x11e7
fmt.Sprintf(0x89b1e7, 0x3, 0xc420395828, 0x1, 0x1, 0xc420395988, 0xc4203d4b40)
	/usr/local/go/src/fmt/print.go:196 +0x6a
log.Printf(0x89b1e7, 0x3, 0xc420395828, 0x1, 0x1)
	/usr/local/go/src/log/log.go:291 +0x53
main.register(0xa87ac0, 0xc4201d8fc0, 0xc4504ab600)
	/home/javdaniel/go/src/github.com/djavorszky/ddn/server/handlers.go:417 +0x413
net/http.HandlerFunc.ServeHTTP(0x8c1c78, 0xa87ac0, 0xc4201d8fc0, 0xc4504ab600)
	/usr/local/go/src/net/http/server.go:1942 +0x44
github.com/djavorszky/ddn/common/srv.Logger.func1(0xa87ac0, 0xc4201d8fc0, 0xc4504ab600)
	/home/javdaniel/go/src/github.com/djavorszky/ddn/common/srv/srv.go:14 +0x9c
net/http.HandlerFunc.ServeHTTP(0xc420132140, 0xa87ac0, 0xc4201d8fc0, 0xc4504ab600)
	/usr/local/go/src/net/http/server.go:1942 +0x44
github.com/gorilla/mux.(*Router).ServeHTTP(0xc4201160a0, 0xa87ac0, 0xc4201d8fc0, 0xc4504ab600)
	/home/javdaniel/go/src/github.com/gorilla/mux/mux.go:114 +0x10c
net/http.serverHandler.ServeHTTP(0xc42019c000, 0xa87ac0, 0xc4201d8fc0, 0xc4504ab300)
	/usr/local/go/src/net/http/server.go:2568 +0x92
net/http.(*conn).serve(0xc4200c3540, 0xa88200, 0xc4203f1680)
	/usr/local/go/src/net/http/server.go:1825 +0x612
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2668 +0x2ce

goroutine 830163 [IO wait, 25 minutes]:
net.runtime_pollWait(0x7fea0039d9d8, 0x72, 0x8)
	/usr/local/go/src/runtime/netpoll.go:164 +0x59
net.(*pollDesc).wait(0xc42012c8b8, 0x72, 0xa84900, 0xa805f8)
	/usr/local/go/src/net/fd_poll_runtime.go:75 +0x38
net.(*pollDesc).waitRead(0xc42012c8b8, 0xc420276000, 0x1000)
	/usr/local/go/src/net/fd_poll_runtime.go:80 +0x34
net.(*netFD).Read(0xc42012c850, 0xc420276000, 0x1000, 0x1000, 0x0, 0xa84900, 0xa805f8)
	/usr/local/go/src/net/fd_unix.go:250 +0x1b7
net.(*conn).Read(0xc4201280c0, 0xc420276000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:181 +0x70
net/http.(*connReader).Read(0xc4203a5700, 0xc420276000, 0x1000, 0x1000, 0x45d172, 0x2, 0xc4201341d0)
	/usr/local/go/src/net/http/server.go:754 +0x140
bufio.(*Reader).fill(0xc420144420)
	/usr/local/go/src/bufio/bufio.go:97 +0x117
bufio.(*Reader).ReadSlice(0xc420144420, 0xa, 0x2, 0x0, 0x0, 0x1, 0x521e77)
	/usr/local/go/src/bufio/bufio.go:338 +0xbb
bufio.(*Reader).ReadLine(0xc420144420, 0xc420114900, 0x100, 0xf8, 0x890720, 0xa, 0x45d172)
	/usr/local/go/src/bufio/bufio.go:367 +0x37
net/textproto.(*Reader).readLineSlice(0xc42023a630, 0xc420473ac8, 0xc420473ac8, 0x410b98, 0x100, 0x890720)
	/usr/local/go/src/net/textproto/reader.go:55 +0x5f
net/textproto.(*Reader).ReadLine(0xc42023a630, 0xc420114900, 0x72, 0x8000000000000000, 0x0)
	/usr/local/go/src/net/textproto/reader.go:36 +0x2f
net/http.readRequest(0xc420144420, 0x0, 0xc420114900, 0x0, 0x0)
	/usr/local/go/src/net/http/request.go:918 +0xa5
net/http.(*conn).readRequest(0xc4200c3180, 0xa88200, 0xc4203a56c0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:934 +0x213
net/http.(*conn).serve(0xc4200c3180, 0xa88200, 0xc4203a56c0)
	/usr/local/go/src/net/http/server.go:1763 +0x49a
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2668 +0x2ce
```

**Comment by djavorszky** [comment]
Issue comes from the fact that the "registry" is not concurrency-safe. Solution is to extract the registry to its own package and make it concurrency safe by introducing sync.Mutex. - see #88

**Comment by djavorszky** [comment]
Fixed data race with fefebc4bd0a49fdf4e103945906fa42dfa85bb3f by introducing a RWMutex
